app/models/concerns/omniauthable.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+module Omniauthable
+  extend ActiveSupport::Concern
+
+  TEMP_EMAIL_PREFIX = 'change@me'
+  TEMP_EMAIL_REGEX = /\Achange@me/
+
+  included do
+    def omniauth_providers
+      providers = []
+      providers << :cas if ENV['CAS_ENABLED']
+      providers
+    end
+
+    def email_verified?
+      email && email !~ TEMP_EMAIL_REGEX
+    end
+  end
+
+  class_methods do
+    def find_for_oauth(auth, signed_in_resource = nil)
+      # EOLE-SSO Patch
+      auth.uid = (auth.uid[0][:uid] || auth.uid[0][:user]) if auth.uid.is_a? Hashie::Array
+
+      # Get the identity and user if they exist
+      identity = Identity.find_for_oauth(auth)
+
+      # If a signed_in_resource is provided it always overrides the existing user
+      # to prevent the identity being locked with accidentally created accounts.
+      # Note that this may leave zombie accounts (with no associated identity) which
+      # can be cleaned up at a later date.
+      user = signed_in_resource ? signed_in_resource : identity.user
+
+      # Create the user if needed
+      user = find_or_create_user_for_oauth(auth) if user.nil?
+
+      # Associate the identity with the user if needed
+      if identity.user != user
+        identity.user = user
+        identity.save!
+      end
+
+      user
+    end
+
+    def find_or_create_user_for_oauth(auth)
+      # Get the existing user by email if the provider gives us a verified email.
+      # If no verified email was provided we assign a temporary email and ask the
+      # user to verify it on the next step via UsersController.finish_signup
+      email_is_verified = auth.info.email && (auth.info.verified || auth.info.verified_email)
+      email = auth.info.email if email_is_verified
+      uid = auth.uid
+      user = Account.find_by(username: uid).try(:user) if uid
```

Comment:
I understand your point. In my opinion, it is the responsibility of the admins to trust or not a specific provider. And the uid can be anything you decide from the provider response :

    cas_options[:uid_field] = 'user'

It's not really a problem with CAS or SAML (for the usecases i can think of) since the user profile cannot be managed by the user himself. 
But for public providers like Facebook, I think you're right, it can be a problem. I like the fact to be able to get the same identity from different providers. But this requires an `Identity` model and the `find_by_uid` mechanism, I cannot think of any other way to do it :/ I'm opened to suggestions.
Another approach would be to implement different callbacks later on, when you decide to implement Facebook oauth or else, with double verification "uid" + "email" ? (cannot be done for CAS since the "email" is not always present)
@Gargron ? @nightpool ? What do you guys think ?